Fix podman network backend fallback on RHEL 9 (HMS-8783)

thozza opened the pull request:
RHEL 9 Podman falls back to the legacy 'cni' network backend when it finds pre-existing container images in storage, interpreting them as a migration from an older version. Since the 'cni' packages are not installed by default, this silently breaks networking for images with embedded containers. Fix this by writing a `defaultNetworkBackend` file into container storage during image build, forcing Podman to use 'netavark'. Add host-config checks to detect this inconsistency and verify container embedding at boot time.

## Architectural Changes

Introduce a `container.NetworkBackend` type and a `GenDefaultNetworkBackendFile` helper in `pkg/container` to generate the `/var/lib/containers/storage/defaultNetworkBackend` file. This file is a Podman-native mechanism that overrides backend auto-detection. The helper accepts a custom storage path to support OSTree images, which relocate container storage to `/usr/share/containers/storage`. The new `PodmanDefaultNetBackend` field on `ImageConfig` drives this behavior declaratively from the distro definition YAML.

## Key Changes

- Add `GenDefaultNetworkBackendFile` in `pkg/container/podman.go` to produce the backend override file, supporting both default and relocated storage paths
- Add `PodmanDefaultNetBackend` option to `ImageConfig` and wire it into `osCustomizations`, gated on both the option being set and containers being present
- Set `podman_default_net_backend: "netavark"` in the RHEL 9 distro definition; Fedora and RHEL 10 are intentionally left unset since they don't have the fallback bug
- Add `container-embedding` host-config check that verifies blueprint containers are present in booted image podman storage
- Add `podman-network-backend` host-config check that detects rootful/rootless backend mismatches

## Breaking Changes

This PR is fully backward compatible.

## Testing

- Table-driven unit tests for `GenDefaultNetworkBackendFile` covering netavark/cni backends and default/custom storage paths
- `osCustomizations` tests verifying the backend file is generated only when both containers and the option are present, including the OSTree relocated-path case
- Distro-level cross-checks asserting RHEL 9 has the backend set to netavark while Fedora and RHEL 10 leave it nil
- Host-config checks tested with mock exec, covering container name matching (including podman's `localhost/` normalization), backend consistency, and error paths

achilleas-k reviewed (dismissed):
Really nice!  Thank.

The last commit message is a bit too long.  Can you make it shorter?

thozza commented:
> Really nice! Thank.
> 
> The last commit message is a bit too long. Can you make it shorter?

Will do.

I hoped that GH will mark this PR as a Draft. :innocent:  My plan is first to see where the consistency check fails, before pushing any fix for RHEL-9.

achilleas-k commented:
Do we still want this?

thozza committed:
pkg/container: add podman default network backend file generation

Certain versions of Podman (notably on RHEL 9) fall back to the legacy
'cni' network backend when they find existing container images in the
system storage, assuming a migration from an older version. This is
problematic for disk images that embed containers as a customization,
because the legacy backend packages are not installed by default.

Add a NetworkBackend type and a helper to generate the
/var/lib/containers/storage/defaultNetworkBackend file, which tells
Podman which backend to use and prevents the unwanted fallback.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
pkg/container: add podman default network backend file generation

Certain versions of Podman (notably on RHEL 9) fall back to the legacy
'cni' network backend when they find existing container images in the
system storage, assuming a migration from an older version. This is
problematic for disk images that embed containers as a customization,
because the legacy backend packages are not installed by default.

Add a NetworkBackend type and a helper to generate the
/var/lib/containers/storage/defaultNetworkBackend file, which tells
Podman which backend to use and prevents the unwanted fallback.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
pkg/distro: add PodmanDefaultNetBackend to ImageConfig and wire it up

Add a new ImageConfig option that specifies the default network backend
for Podman. When set and the image embeds container images, the value is
written to /var/lib/containers/storage/defaultNetworkBackend during
image build.

This prevents Podman from falling back to the legacy 'cni' backend when
it finds pre-existing container images in storage, which it interprets
as a system migration.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
pkg/distro: add PodmanDefaultNetBackend to ImageConfig and wire it up

Add a new ImageConfig option that specifies the default network backend
for Podman. When set and the image embeds container images, the value is
written to /var/lib/containers/storage/defaultNetworkBackend during
image build.

This prevents Podman from falling back to the legacy 'cni' backend when
it finds pre-existing container images in storage, which it interprets
as a system migration.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
distro/rhel-9: set podman default network backend to netavark

RHEL 9 Podman falls back to the legacy 'cni' network backend when it
finds pre-existing container images in storage, but the 'cni' packages
are not installed by default. Force 'netavark' so that images with
embedded containers work out of the box.

This only affects RHEL 9 / CentOS Stream 9; newer distros (and newer
podman versions) don't have this fallback logic.

Regenerate test manifests. All el9 / c9s manifests that embed
containers now get the podman default network backend set.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
distro/rhel-9: set podman default network backend to netavark

RHEL 9 Podman falls back to the legacy 'cni' network backend when it
finds pre-existing container images in storage, but the 'cni' packages
are not installed by default. Force 'netavark' so that images with
embedded containers work out of the box.

This only affects RHEL 9 / CentOS Stream 9; newer distros (and newer
podman versions) don't have this fallback logic.

Regenerate test manifests. All el9 / c9s manifests that embed
containers now get the podman default network backend set.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
distro: add tests for PodmanDefaultNetBackend

Verify the podman default network backend behavior at multiple levels:
- YAML loading: fake distro YAML with podman_default_net_backend loads
  correctly into ImageConfig
- osCustomizations: the backend file is generated only when both
  containers are present and the option is set
- Real distro cross-check: RHEL 9 has it set to netavark, while
  Fedora and RHEL 10 leave it unset

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
distro: add tests for PodmanDefaultNetBackend

Verify the podman default network backend behavior at multiple levels:
- YAML loading: fake distro YAML with podman_default_net_backend loads
  correctly into ImageConfig
- osCustomizations: the backend file is generated only when both
  containers are present and the option is set
- Real distro cross-check: RHEL 9 has it set to netavark, while
  Fedora and RHEL 10 leave it unset

Also move the backend file generation in osCustomizations() to after
the file customizations assignment, which would otherwise overwrite it.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
cmd/check-host-config: add container embedding check

Verify that containers listed in the blueprint are actually present in
the booted image's podman storage.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
cmd/check-host-config: add container embedding check

Verify that containers listed in the blueprint are actually present in
the booted image's podman storage.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
cmd/check-host-config: add podman network backend consistency check

Verify that rootful and rootless podman report the same network
backend. When containers are embedded as root into the image (the
default behavior), some podman versions interpret the existing storage
as a migration and fall back to 'cni' for rootful only, leaving
rootless on 'netavark'. In practice, the desired behavior is that podman
uses the same network backend, regardless if there is an embedded
container or not.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
cmd/check-host-config: add podman network backend consistency check

Verify that rootful and rootless podman report the same network
backend. When containers are embedded as root into the image (the
default behavior), some podman versions interpret the existing storage
as a migration and fall back to 'cni' for rootful only, leaving
rootless on 'netavark'. In practice, the desired behavior is that podman
uses the same network backend, regardless if there is an embedded
container or not.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
test/manifests: embed container in all-customizations config

Add podman to the package list and a container reference to the
all-customizations test blueprint so that the container embedding and
network backend consistency host checks have something to verify.

Regenerate manifest checksums to reflect the updated test config and
the new defaultNetworkBackend file in RHEL 9 / CentOS Stream 9
manifests.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
Schutzfile: bump rngseed to force full rebuild

Make sure that the newly added checks are run on all images.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

brlane-rht commented on cmd/check-host-config/check/podman_network_backend_test.go:
```diff
@@ -0,0 +1,103 @@
+				},
+			},
+			wantErr: check.ErrCheckFailed,
+		},
```
This should also test the empty backend string / unknown cases.
Also, what's the expected result if both are unknown? Should that be a pass or a fail?
